### PR TITLE
Fuse decorator without paren

### PR DESCRIPTION
**Pull request by niboshi**
Retarget of this PR: https://github.com/pfnet/chainer/pull/2548

Fixes https://github.com/pfnet/chainer/issues/2547

### CODE REVIEW

**Comment by jekbradbury** [comment]
It would be nice if there were a `@fuse(cached=True)` or `@fuse_cached` which enables the kernel generated by one call to the function to be reused for another call.

File: cupy/core/fusion.py [hunk 1]
```diff
@@ -649,7 +650,15 @@ def fuse(input_num=None, reduce=None, post_map=lambda x: x):
             If not assigned, post_map step is skipped.
     """
     util.experimental('cupy.core.fusion')
-    return lambda f: Fusion(f, input_num, reduce, post_map)
+
+    def wrapper(f, input_num=None, reduce=None, post_map=lambda x: x):
+        return Fusion(f, input_num, reduce, post_map)
+
+    if len(args) == 1 and len(kwargs) == 0 and callable(args[0]):
+        return functools.update_wrapper(wrapper(args[0]), args[0])
```
Comment (takagi) [inline cupy/core/fusion.py]: `functools.wraps` is more convenient and is used in other places of CuPy.

File: cupy/core/fusion.py [hunk 2]
```diff
@@ -649,7 +650,15 @@ def fuse(input_num=None, reduce=None, post_map=lambda x: x):
...
+    else:
+        return lambda f: functools.update_wrapper(
```
Comment (takagi) [inline cupy/core/fusion.py]: ditto(`functools.wraps`)

File: cupy/core/fusion.py (hunk 1 above)
Comment (niboshi) [inline cupy/core/fusion.py]: I think `functools.wraps` cannot be used here as it is a decorator, and what we want to look like `args[0]` is a `Fusion` instance, not `wrapper`.

File: cupy/core/fusion.py (hunk 2 above)
Comment (niboshi) [inline cupy/core/fusion.py]: ditto.

File: cupy/core/fusion.py (hunk 1 above)
Comment (takagi) [inline cupy/core/fusion.py]: Ah, now I got it.

**Comment by takagi** [comment]
LGTM!

**Pull request merged**